Client/Frontend/Home/TopSitesPanel.swift

```diff
@@ -204,15 +204,8 @@ private class TopSitesDataSource: NSObject, UICollectionViewDataSource {
         if indexPath.item < 6 {
             let cell = collectionView.dequeueReusableCellWithReuseIdentifier(ThumbnailIdentifier, forIndexPath: indexPath) as ThumbnailCell
             cell.textLabel.text = site.title.isEmpty ? site.url : site.title
-            cell.image = nil
-
-            if let url = NSURL(string: site.url) {
-                profile.thumbnails.get(url) { (thumbnail: Thumbnail?) in
-                    if let thumbnail = thumbnail {
-                        cell.image = thumbnail.image
-                    }
-                }
-            }
+            let key = SDWebThumbnails.getKey(site.url)
+            cell.imageView.sd_setImageWithURL(NSURL(string: key))
```

Comment:
Note that you *cannot* set `cell.imageView` asynchronously. The reason is that the cells are reused/recycled. So there is a change that when `thumbnails.get()` finishes, it is setting `cell.image` for a cell that is not the same cell as you started with.

It currently only works because our collectionview is pretty simple. (But I think we have this same issue for table views where we show favicons)

It is odd that SDWebImage shows this exact usage on their Github project homepage as an example, because it is most definitely a bug to do that.

The correct way to do this is to have some check inside the `thumbnails.get()` completion block to find out if the cell is actually still the right one. For example by comparing the URL of the thumbnail. Another approach that you often see is using notifications to notify a cell that its image has become available in the cache. The cell then registers/unregisters for these notifications when it is being added/removed to/from the collection view.